relay/cached_accessor.go

```diff
@@ -0,0 +1,118 @@
+package relay
+
+import (
+	lru "github.com/hashicorp/golang-lru/v2"
+	"sync"
+)
+
+// CachedAccessor is an interface for accessing a resource that is cached. It assumes that cache misses
+// are expensive, and prevents multiple concurrent cache misses for the same key.
+type CachedAccessor[K comparable, V any] interface {
+	// Get returns the value for the given key. If the value is not in the cache, it will be fetched using the Accessor.
+	Get(key K) (*V, error)
+}
+
+// Accessor is function capable of fetching a value from a resource. Used by CachedAccessor when there is a cache miss.
+type Accessor[K comparable, V any] func(key K) (V, error)
+
+// accessResult is a struct that holds the result of an Accessor call.
+type accessResult[V any] struct {
+	// wg.Wait() will block until the value is fetched.
+	wg sync.WaitGroup
+	// value is the value fetched by the Accessor, or nil if there was an error.
+	value *V
+	// err is the error returned by the Accessor, or nil if the fetch was successful.
+	err error
+}
+
+var _ CachedAccessor[string, string] = &cachedAccessor[string, string]{}
+
+// Future work: the cache used in this implementation is suboptimal when storing items that have a large
+// variance in size. The current implementation uses a fixed size cache, which requires the cached to be
+// sized to the largest item that will be stored. This cache should be replaced with an implementation
+// whose size can be specified by memory footprint in bytes.
+
+// cachedAccessor is an implementation of CachedAccessor.
+type cachedAccessor[K comparable, V any] struct {
+	// used to synchronize access to the cache and lookupsInProgress
+	lock sync.Mutex
+
+	// lookupsInProgress has an entry for each key that is currently being looked up via the accessor. The value
+	// is written into the channel when it is eventually fetched. If a key is requested more than once while a
+	// lookup in progress, the second (and following) requests will wait for the result of the first lookup
+	// to be written into the channel.
+	lookupsInProgress map[K]*accessResult[V]
```

Comment:
maybe we can use `sync.Map` and ditch the mutex?

Reply:
Done.